clients/unshaded/src/main/java/tachyon/client/netty/ClientHandler.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the University of California, Berkeley under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package tachyon.client.netty;
+
+import java.io.IOException;
+import java.util.HashSet;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+
+import tachyon.Constants;
+import tachyon.network.protocol.RPCBlockResponse;
+import tachyon.network.protocol.RPCMessage;
+import tachyon.network.protocol.RPCResponse;
+
+/**
+ * This handles all the messages received by the client channel.
+ */
+@ChannelHandler.Sharable
+public final class ClientHandler extends SimpleChannelInboundHandler<RPCMessage> {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  /**
+   * The interface for listeners to implement to receive callbacks when messages are received.
+   */
+  public interface ResponseListener {
+    /** This method will be called when a message is received on the client. */
+    void onResponseReceived(RPCResponse response);
+  }
+
+  private final HashSet<ResponseListener> mListeners;
+
+  public ClientHandler() {
+    mListeners = new HashSet<ResponseListener>();
+  }
+
+  public void addListener(ResponseListener listener) {
+    mListeners.add(listener);
+  }
+
+  public void removeListener(ResponseListener listener) {
+    mListeners.remove(listener);
+  }
+
+  @Override
+  public void channelRead0(final ChannelHandlerContext ctx, final RPCMessage msg)
+      throws IOException {
+    switch (msg.getType()) {
+      case RPC_BLOCK_RESPONSE:
+        handleBlockResponse(ctx, (RPCBlockResponse) msg);
+        break;
+      default:
+        throw new IllegalArgumentException("No handler implementation for rpc response type: "
+            + msg.getType());
```

Comment:
Is it hardcoded which message types get forwarded to the listeners? In the future will the listeners be able to selectively listen or do all messages get forwarded and then the listeners need to filter in their implementation?